src/BlockDiagonals.jl

```diff
@@ -0,0 +1,340 @@
+    size(A, 2) == size(B, 1) || throw(DimensionMismatch(
+        "A has dimensions $(size(A)) but B has dimensions $(size(B))"
+    ))
+end
+
+function Base.:*(B::BlockDiagonal, M::AbstractMatrix)
+    _check_matmul_dims(B, M)
+    st = 1  # start
+    ed = 0  # end
+    d = Matrix[]
+    for block in blocks(B)
+        ed += size(block, 2)
+        push!(d, block * M[st:ed, :])
+        st = ed + 1
+    end
+    return reduce(vcat, d)::Matrix
+end
+
+function Base.:*(M::AbstractMatrix, B::BlockDiagonal)
+    _check_matmul_dims(M, B)
+    st = 1  # start
+    ed = 0  # end
+    d = Matrix[]
+    for block in blocks(B)
+        ed += size(block, 1)
+        push!(d, M[:, st:ed] * block)
+        st = ed + 1
+    end
+    return reduce(hcat, d)::Matrix
+end
+
+# Diagonal
+function Base.:*(B::BlockDiagonal, M::Diagonal)::BlockDiagonal
+    _check_matmul_dims(B, M)
+    A = copy(B)
+    d = diag(M)
+    col = 1
+    for (p, block) in enumerate(blocks(B))
+        ncols = size(block, 2)
+        cols = col:(col + ncols-1)
+        for (j, c) in enumerate(cols)
+            A[Block(p)][:, j] *= d[c]
+        end
+        col += ncols
+    end
+    return A
+end
+
+function Base.:*(M::Diagonal, B::BlockDiagonal)::BlockDiagonal
+    _check_matmul_dims(M, B)
+    A = copy(B)
+    d = diag(M)
+    row = 1
+    for (p, block) in enumerate(blocks(B))
+        nrows = size(block, 1)
+        rows = row:(row + nrows-1)
+        for (i, r) in enumerate(rows)
+            A[Block(p)][i, :] *= d[r]
+        end
+        row += nrows
+    end
+    return A
+end
+
+## Division
+Base.:/(B::BlockDiagonal, n::Number) = BlockDiagonal(blocks(B) ./ n)
+
+## LinearAlgebra
+for f in (:adjoint, :eigvecs, :inv, :transpose)
+    @eval LinearAlgebra.$f(B::BlockDiagonal) = BlockDiagonal(map($f, blocks(B)))
+end
+
+LinearAlgebra.diag(B::BlockDiagonal) = mapreduce(diag, vcat, blocks(B))
+LinearAlgebra.det(B::BlockDiagonal) = prod(det, blocks(B))
+LinearAlgebra.logdet(B::BlockDiagonal) = sum(logdet, blocks(B))
+LinearAlgebra.tr(B::BlockDiagonal) = sum(tr, blocks(B))
+
+# Real matrices can have Complex eigenvalues; `eigvals` is not type stable.
+function LinearAlgebra.eigvals(B::BlockDiagonal)
+    eigs = mapreduce(eigvals, vcat, blocks(B))
+    # Currently no convention for sorting Complex eigenvalues.
+    # This may change in later a Julia version https://github.com/JuliaLang/julia/pull/21598
+    eigs isa Vector{<:Complex} && return eigs
+    return sort!(eigs)
+end
+
+svdvals_blockwise(B::BlockDiagonal) = mapreduce(svdvals, vcat, blocks(B))
+LinearAlgebra.svdvals(B::BlockDiagonal) = sort!(svdvals_blockwise(B); rev=true)
+
+# `B = U * Diagonal(S) * Vt` with `U` and `Vt` `BlockDiagonal` (`S` only sorted block-wise).
+function svd_blockwise(B::BlockDiagonal; full::Bool=false)
+    Fs = svd.(blocks(B); full=full)
+    U = BlockDiagonal([F.U for F in Fs])
+    S = mapreduce(F -> F.S, vcat, Fs)
+    Vt = BlockDiagonal([F.Vt for F in Fs])
+    return U, S, Vt
+end
```

Comment:
I'd be interested to see whether something like this is faster, as it only makes one pass over the blocks:
```julia
function svd_blockwise(B::BlockDiagonal{T}; full::Bool=false) where T
    U = Matrix{T}[]
    S = Vector{T}[]
    Vt = Matrix{T}[]
    for b in blocks(B)
        F = svd(b, full=full)
        push!(U, F.U)
        push!(S, F.S)
        push!(Vt, F.Vt)
    end
    BlockDiagonal(U), S, BlockDiagonal(Vt)
end
```

Reply:
Nice! Yes, this scales better
```julia
julia> using BenchmarkTools, BlockDiagonals, LinearAlgebra

julia> function svd_blockwise1(B::BlockDiagonal; full::Bool=false)
           Fs = svd.(blocks(B); full=full)
           U = BlockDiagonal([F.U for F in Fs])
           S = mapreduce(F -> F.S, vcat, Fs)
           Vt = BlockDiagonal([F.Vt for F in Fs])
           return U, S, Vt
       end
svd_blockwise1 (generic function with 1 method)

julia> function svd_blockwise2(B::BlockDiagonal{T}; full::Bool=false) where T
           U = Matrix{T}[]
           S = Vector{T}()
           Vt = Matrix{T}[]
           for b in blocks(B)
               F = svd(b, full=full)
               push!(U, F.U)
               append!(S, F.S)
               push!(Vt, F.Vt)
           end
           BlockDiagonal(U), S, BlockDiagonal(Vt)
       end
svd_blockwise2 (generic function with 1 method)

julia> B = BlockDiagonal(repeat([randn(2,2), rand(3,3), randn(4,4)], 1));

julia> @btime svd_blockwise1($B);
  11.386 μs (65 allocations: 10.47 KiB)

julia> @btime svd_blockwise2($B);
  11.978 μs (66 allocations: 10.42 KiB)

julia> B = BlockDiagonal(repeat([randn(2,2), rand(3,3), randn(4,4)], 10));

julia> @btime svd_blockwise1($B);
  103.529 μs (499 allocations: 108.73 KiB)

julia> @btime svd_blockwise2($B);
  100.536 μs (482 allocations: 97.64 KiB)

julia> B = BlockDiagonal(repeat([randn(2,2), rand(3,3), randn(4,4)], 100));

julia> @btime svd_blockwise1($B);
  1.071 ms (4825 allocations: 2.00 MiB)

julia> @btime svd_blockwise2($B);
  979.620 μs (4549 allocations: 971.83 KiB)
```

Reply:
although `Matrix{T}` will need to be `Matrix{float(T)}` (cos `svd` on `BlockDiagonal{Int}` will return floats)